File: web/src/lib/files-api.ts
// ── Files client ─────────────────────────────────────────────────────────────
//
// One directory listing, annotated with git status, for Mission Control's Files
// applet. The server half is internal/server/files_api.go and its contract is
// the comment at the top of that file; this module is the browser's half of it.
//
// TWO PROPERTIES OF THE WIRE SHAPE DRIVE EVERYTHING BELOW:
//
//   1. `entries` is never null, and `gitAvailable:false` is a DEGRADED LISTING,
//      not an error. A directory outside any worktree still returns every file
//      it holds, with one sentence in `gitError` saying why the annotations are
//      missing. Losing the git half must never cost the caller its listing, so
//      nothing here throws on it.
//
//   2. A 4xx/5xx carries `{"error": "<sentence>"}` written by a human for a
//      human -- "/nope does not exist", "\"web/src\" is not an absolute path".
//      fetchFiles rethrows that sentence verbatim (update.ts:95-105's idiom) so
//      the applet can show the server's own words instead of "HTTP 404".
//
// Every read goes through a hand-written narrowing function with a default per
// field. `await res.json()` is `any`, and one bare cast would put an undefined
// where the applet expects a string for the rest of the app's life.

import { apiPath } from './base-path.js';

/**
 * What git says about one entry. '' is the ABSENCE of a status, not a status:
 * an unchanged file, or any file at all when git could not be consulted.
 *
 * Mirrors the fileStatus* constants in internal/server/files_api.go. A
 * directory reports 'modified' when anything under it changed, whatever that
 * change was -- see statusFor() there for why the roll-up cannot be more
 * specific than that without lying.
 */
export type FileStatus =
  | ''
  | 'modified'
  | 'added'
  | 'deleted'
  | 'untracked'
  | 'renamed'
  | 'conflicted';

const FILE_STATUSES: readonly string[] = [
  'modified',
  'added',
  'deleted',
  'untracked',
  'renamed',
  'conflicted',
];

/** One row of a listing. `size`/`modified` come from Lstat, so a symlink
 * reports its own size and `dir:false` even when it points at a directory. */
export interface FileEntry {
  name: string;
  dir: boolean;
  /** Bytes. */
  size: number;
  /** Unix seconds. */
  modified: number;
  status: FileStatus;
}

/** GET /api/files. */
export interface FilesListing {
  /** The cleaned absolute directory that was listed. */
  path: string;
  /** The directory above `path`, or '' at the filesystem root. */
  parent: string;
  /** '' when `path` is not inside a worktree. */
  repoRoot: string;
  /** '' when detached, unborn, or not a worktree. */
  branch: string;
  /** false when git is missing, this is no worktree, or status failed. */
  gitAvailable: boolean;
  /** One human sentence when `gitAvailable` is false; '' otherwise. */
  gitError: string;
  /** Directories first, then files, each case-insensitive. Never null. */
  entries: FileEntry[];
  /** Write availability for precisely this rendered local directory. */
  upload: FilesUploadAvailability;
}

/** A local, server-scoped upload binding; no host path or capability is exposed. */
export interface FilesUploadAvailability {
  available: boolean;
  /** A safe, user-actionable reason when upload is unavailable. */
  reason: string;
}

/** The only browser-visible outcomes of one streamed file upload. */
export interface FilesUploadResult {
  status: 'complete' | 'conflict' | 'failed' | 'cancelled';
  name: string;
  size: number;
  message: string;
}

export type FilesUploadResolution = 'new' | 'keep' | 'replace';

function parseFileStatus(raw: unknown): FileStatus {
  return typeof raw === 'string' && FILE_STATUSES.includes(raw) ? (raw as FileStatus) : '';
}

/** Narrow one untrusted row, defaulting anything unexpected. */
export function parseFileEntry(raw: unknown): FileEntry {
  if (raw === null || typeof raw !== 'object' || Array.isArray(raw)) {
    return { name: '', dir: false, size: 0, modified: 0, status: '' };
  }
  const r = raw as Record<string, unknown>;
  return {
    name: typeof r['name'] === 'string' ? r['name'] : '',
    dir: r['dir'] === true,
    size: typeof r['size'] === 'number' && Number.isFinite(r['size']) ? r['size'] : 0,
    modified:
      typeof r['modified'] === 'number' && Number.isFinite(r['modified']) ? r['modified'] : 0,
    status: parseFileStatus(r['status']),
  };
}

/**
 * Narrow untrusted JSON into a FilesListing.
 *
 * A row with no `name` is dropped rather than rendered: it can only come from a
 * response that is not this endpoint's, and a blank row in a file list is worse
 * than a short list.
 */
export function parseFilesListing(raw: unknown): FilesListing {
  if (raw === null || typeof raw !== 'object' || Array.isArray(raw)) {
    return {
      path: '',
      parent: '',
      repoRoot: '',
      branch: '',
      gitAvailable: false,
      gitError: '',
      entries: [],
      upload: { available: false, reason: '' },
    };
  }
  const r = raw as Record<string, unknown>;
  const entries = Array.isArray(r['entries'])
    ? (r['entries'] as unknown[]).map(parseFileEntry).filter((e) => e.name !== '')
    : [];
  return {
    path: typeof r['path'] === 'string' ? r['path'] : '',
    parent: typeof r['parent'] === 'string' ? r['parent'] : '',
    repoRoot: typeof r['repoRoot'] === 'string' ? r['repoRoot'] : '',
    branch: typeof r['branch'] === 'string' ? r['branch'] : '',
    gitAvailable: r['gitAvailable'] === true,
    gitError: typeof r['gitError'] === 'string' ? r['gitError'] : '',
    entries,
    upload: parseFilesUploadAvailability(r['upload']),
  };
}

function parseFilesUploadAvailability(raw: unknown): FilesUploadAvailability {
  if (raw === null || typeof raw !== 'object' || Array.isArray(raw)) {
    return { available: false, reason: '' };
  }
  const r = raw as Record<string, unknown>;
  return {
    available: r['available'] === true,
    reason: typeof r['reason'] === 'string' ? r['reason'] : '',
  };
}

function parseFilesUploadResult(raw: unknown): FilesUploadResult {
  const r = raw !== null && typeof raw === 'object' && !Array.isArray(raw) ? raw as Record<string, unknown> : {};
  const status = r['status'];
  return {
    status:
      status === 'complete' || status === 'conflict' || status === 'cancelled'
        ? status
        : 'failed',
    name: typeof r['name'] === 'string' ? r['name'] : '',
    size: typeof r['size'] === 'number' && Number.isFinite(r['size']) ? r['size'] : 0,
    message: typeof r['message'] === 'string' ? r['message'] : '',
  };
}

/**
 * GET /api/files[?path=<absolute path>].
 *
 * `path` omitted means "wherever this server process is" -- the only sensible
 * opening directory for a browser that has not chosen one yet, and the ONLY way
 * to learn the server's cwd. Passing a relative path is a caller bug and the
 * server says so in words.
 *
 * `signal` is how the applet honours the inactive rule: an applet that goes
 * inactive mid-flight aborts, and an aborted fetch rejects with an AbortError
 * the caller is expected to ignore rather than render.
 *
 * Throws the server's own `error` sentence on any non-2xx, so the applet can
 * show "/gone does not exist" instead of "HTTP 404".
 */
export async function fetchFiles(path?: string, signal?: AbortSignal): Promise<FilesListing> {
  const url =
    path === undefined
      ? apiPath('/api/files')
      : `${apiPath('/api/files')}?${new URLSearchParams({ path }).toString()}`;
  const res = await fetch(url, signal ? { signal } : undefined);
  const body = (await res.json().catch(() => ({}))) as Record<string, unknown>;
  if (!res.ok) {
    const err =
      typeof body['error'] === 'string' && body['error'] !== ''
        ? body['error']
        : `fetchFiles: HTTP ${res.status}`;
    throw new Error(err);
  }
  return parseFilesListing(body);
}

/**
 * Stream one browser File through the scoped Files endpoint.
 *
 * XMLHttpRequest is intentional here. Fetch does not provide upload progress;
 * XHR streams FormData without materializing file bytes in JS and supplies the
 * progress/cancellation semantics the in-applet queue needs.
 */
export function uploadFilesFile(
  file: File,
  resolution: FilesUploadResolution,
  signal: AbortSignal,
  onProgress: (loaded: number, total: number) => void,
): Promise<FilesUploadResult> {
  return new Promise((resolve, reject) => {
    const xhr = new XMLHttpRequest();
    xhr.open('POST', apiPath('/api/files/upload'));
    xhr.responseType = 'json';
    xhr.withCredentials = true;
    xhr.setRequestHeader('X-Muxterm-Files-Upload', '1');
    xhr.setRequestHeader('X-Muxterm-Files-Resolution', resolution);
    xhr.upload.onprogress = (event) => {
      if (event.lengthComputable) onProgress(event.loaded, event.total);
    };
    xhr.onerror = () => reject(new Error('The upload could not reach muxterm. No file was saved.'));
    xhr.onabort = () => reject(new DOMException('Upload cancelled.', 'AbortError'));
    xhr.onload = () => {
      const raw = xhr.response ?? (() => {
        try {
          return JSON.parse(xhr.responseText) as unknown;
        } catch {
          return {};
        }
      })();
      const result = parseFilesUploadResult(raw);
      if ((xhr.status >= 200 && xhr.status < 300) || xhr.status === 409 || xhr.status === 499) {
        resolve(result);
        return;
      }
      reject(new Error(result.message || 'The upload could not be completed. No file was saved.'));
    };
    const abort = () => xhr.abort();
    signal.addEventListener('abort', abort, { once: true });
    const form = new FormData();
    form.append('file', file, file.name);
    xhr.send(form);
  });
}

File: internal/server/files_api.go
package server

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"net/http"
	"os"
	"os/exec"
	"path/filepath"
	"sort"
	"strings"
	"time"
)

// The /api/files route: one directory listing, annotated with git status, for
// Mission Control's Files applet.
//
//	GET /api/files[?path=<absolute path>]   list that directory
//
// AuthMiddleware protects this route at mux registration, exactly like the
// config, AI, tunnel and remotes routes.
//
// SECURITY. This route is wrapped in protect() like every other /api route, and
// it adds NO authority beyond what /ws already grants -- the same auth boundary
// hands out a PTY, and a shell can cat any file this handler can list. So the
// guard here is CORRECTNESS (absolute, cleaned, actually a directory) rather
// than a jail, and it is deliberately not a jail. Do not invent a chroot: it
// would constrain the file browser without constraining the terminal sitting
// next to it, which is security theatre with a maintenance cost.
//
// The git half is best-effort by design. A directory outside any worktree, a
// machine with no git, or a `git status` that fails still returns the full
// listing with gitAvailable=false and one sentence in gitError. Losing the
// annotations must never cost the user the file listing.

// filesGitDeadline bounds the WHOLE git interrogation of one directory -- the
// toplevel lookup, the branch lookup and the status pass together, not each --
// so a wedged git cannot stack three timeouts onto one GET.
const filesGitDeadline = 5 * time.Second

// The status vocabulary on the wire. An entry with nothing to report carries
// "", which is why there is no constant for it: it is the absence of a status,
// not a status.
const (
	fileStatusModified   = "modified"
	fileStatusAdded      = "added"
	fileStatusDeleted    = "deleted"
	fileStatusUntracked  = "untracked"
	fileStatusRenamed    = "renamed"
	fileStatusConflicted = "conflicted"
)

// fileEntry is one row of the listing.
//
// Size and Modified come from Lstat, NOT Stat: a symlink reports its own size
// and its own mtime, and a symlink to a directory reports dir=false. Following
// it would let a listing hang on a dead NFS mount and would report a directory
// the user cannot expand as expandable.
type fileEntry struct {
	Name     string `json:"name"`
	Dir      bool   `json:"dir"`
	Size     int64  `json:"size"`
	Modified int64  `json:"modified"` // unix seconds
	Status   string `json:"status"`   // "" | modified | added | deleted | untracked | renamed | conflicted
}

// filesListResponse is GET /api/files. Entries is ALWAYS present and never
// null: the browser iterates it unconditionally.
type filesListResponse struct {
	Path         string      `json:"path"`         // the cleaned absolute directory listed
	Parent       string      `json:"parent"`       // "" at the filesystem root
	RepoRoot     string      `json:"repoRoot"`     // "" when not inside a worktree
	Branch       string      `json:"branch"`       // "" when detached or not a worktree
	GitAvailable bool        `json:"gitAvailable"` // false when git is missing, this is no worktree, or status failed
	GitError     string      `json:"gitError"`     // one human sentence when GitAvailable is false
	Entries      []fileEntry `json:"entries"`
}

func writeFilesJSON(w http.ResponseWriter, code int, v any) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(code)
	json.NewEncoder(w).Encode(v) //nolint:errcheck
}

// writeFilesError renders every 4xx/5xx body: {"error": "<sentence>"}.
func writeFilesError(w http.ResponseWriter, code int, err error) {
	writeFilesJSON(w, code, map[string]any{"error": err.Error()})
}

// handleFilesList answers GET /api/files[?path=<absolute path>].
//
// ?path is optional as a KEY and required as a VALUE: omitting it entirely
// means "wherever this server process is", which is the only sensible opening
// directory for a browser that has not chosen one yet; sending it empty is a
// caller bug and says so. Anything relative is refused rather than resolved
// against the server's cwd, because a browser that thinks it is asking for
// "web/src" and gets a listing of somewhere else has been lied to.
//
// AuthMiddleware protects this route at mux registration.
func (s *Server) handleFilesList(w http.ResponseWriter, r *http.Request) {
	q := r.URL.Query()
	dir := q.Get("path")
	if !q.Has("path") {
		cwd, err := os.Getwd()
		if err != nil {
			writeFilesError(w, http.StatusInternalServerError, fmt.Errorf("this server cannot determine its own working directory: %w", err))
			return
		}
		dir = cwd
	}
	if strings.TrimSpace(dir) == "" {
		writeFilesError(w, http.StatusBadRequest, errors.New(`"path" is required and must be an absolute path`))
		return
	}
	if !filepath.IsAbs(dir) {
		writeFilesError(w, http.StatusBadRequest, fmt.Errorf("%q is not an absolute path", dir))
		return
	}
	dir = filepath.Clean(dir)

	info, err := os.Stat(dir)
	switch {
	case os.IsNotExist(err):
		writeFilesError(w, http.StatusNotFound, fmt.Errorf("%s does not exist", dir))
		return
	case err != nil:
		writeFilesError(w, http.StatusInternalServerError, err)
		return
	case !info.IsDir():
		writeFilesError(w, http.StatusBadRequest, fmt.Errorf("%s is not a directory", dir))
		return
	}

	git := inspectGit(r.Context(), dir)

	entries, err := readDirEntries(dir, git.changed)
	if err != nil {
		writeFilesError(w, http.StatusInternalServerError, err)
		return
	}

	writeFilesJSON(w, http.StatusOK, filesListResponse{
		Path:         dir,
		Parent:       parentOf(dir),
		RepoRoot:     git.repoRoot,
		Branch:       git.branch,
		GitAvailable: git.err == "",
		GitError:     git.err,
		Entries:      entries,
	})
}

// parentOf returns the directory above dir, or "" when dir IS the top. The
// filesystem root is its own Dir, and so is a volume root on Windows, which is
// what makes this comparison the portable spelling of "no parent".
func parentOf(dir string) string {
	parent := filepath.Dir(dir)
	if parent == dir {
		return ""
	}
	return parent
}

// readDirEntries lists dir and annotates each row from the changed-path map.
//
// An entry whose Lstat fails is SKIPPED rather than failing the listing: a file
// deleted between the readdir and the stat is normal on a live worktree, and
// losing the whole directory over one racing temp file would make the applet
// flicker into an error on every build.
func readDirEntries(dir string, changed map[string]string) ([]fileEntry, error) {
	names, err := os.ReadDir(dir)
	if err != nil {
		return nil, err
	}

	// Never null. An empty directory is an empty array.
	out := make([]fileEntry, 0, len(names))
	for _, de := range names {
		// Lstat, not Stat: see fileEntry. Dotfiles are included -- in a
		// worktree they are exactly the files a developer edits.
		info, err := os.Lstat(filepath.Join(dir, de.Name()))
		if err != nil {
			continue
		}
		row := fileEntry{
			Name:     de.Name(),
			Dir:      info.IsDir(),
			Size:     info.Size(),
			Modified: info.ModTime().Unix(),
		}
		row.Status = statusFor(filepath.Join(dir, de.Name()), row.Dir, changed)
		out = append(out, row)
	}
	out = appendDeleted(dir, changed, out)

	// Directories first, then files, each case-insensitively by name. The
	// case-sensitive tie-break makes the order total, so two names differing
	// only in case cannot swap places between polls.
	sort.Slice(out, func(i, j int) bool {
		if out[i].Dir != out[j].Dir {
			return out[i].Dir
		}
		li, lj := strings.ToLower(out[i].Name), strings.ToLower(out[j].Name)
		if li != lj {
			return li < lj
		}
		return out[i].Name < out[j].Name
	})
	return out, nil
}

// appendDeleted adds a row for every path git calls DELETED that is a direct
// child of dir and is therefore no longer on disk for ReadDir to have found.
//
// WHY THIS EXISTS. The listing is built from what the filesystem holds, and a
// deleted file is precisely the thing that is not there any more -- so without
// this, a "changed" filter would confidently show a directory with an unstaged
// `rm` in it as having nothing changed. That is the one failure mode a status
// filter must not have: silently omitting a change while claiming to show
// changes. The frontend has always been ready for these rows (a deleted name is
// struck through); it was the server that never produced one.
//
// Only DIRECT children, and only paths that really are absent: a path that
// still exists was already emitted above with its own status, and emitting it
// twice would put the same name in the list twice.
//
// Size and Modified are zero, honestly: there is no file to measure.
func appendDeleted(dir string, changed map[string]string, out []fileEntry) []fileEntry {
	for p, st := range changed {
		if st != fileStatusDeleted {
			continue
		}
		if filepath.Dir(p) != dir {
			continue
		}
		if _, err := os.Lstat(p); err == nil {
			continue
		}
		out = append(out, fileEntry{Name: filepath.Base(p), Status: fileStatusDeleted})
	}
	return out
}

// statusFor resolves one entry's status against the changed-path map.
//
// A file gets a status only when git named it exactly. A DIRECTORY also gets
// "modified" when anything under it changed, whatever that change was: rolling
// up an "untracked" child as "untracked" would claim the directory itself is
// untracked, which is a different and usually false statement. "modified"
// means "there is something in here", which is all a collapsed row can honestly
// say.
//
// The scan is linear in the number of changed paths per entry. Bounded by a
// directory listing on one side and a worktree's dirty set on the other, that
// is a few thousand string comparisons on a bad day -- cheaper than the git
// subprocess that produced the map.
func statusFor(abs string, isDir bool, changed map[string]string) string {
	if st, ok := changed[abs]; ok {
		return st
	}
	if !isDir {
		return ""
	}
	prefix := abs + string(filepath.Separator)
	for p := range changed {
		if strings.HasPrefix(p, prefix) {
			return fileStatusModified
		}
	}
	return ""
}

// filesGit is what one git interrogation of a directory yields. err carries one
// human sentence and is the single source of gitAvailable: when it is empty the
// other three fields are trustworthy, and when it is not they are whatever
// could be learned before the failure.
type filesGit struct {
	repoRoot string
	branch   string
	changed  map[string]string // absolute path -> status token
	err      string
}

// inspectGit derives repo root, branch and the changed-path map for dir.
//
// Every failure path here is survivable and returns a filesGit the caller can
// still render: git missing, dir outside any worktree, or a status that failed.
// The one thing it never does is return an error that costs the caller its
// listing.
func inspectGit(ctx context.Context, dir string) filesGit {
	out := filesGit{changed: map[string]string{}}

	// LookPath first, so a machine with no git is a clean degraded state
	// rather than an exec error dressed up as a git failure.
	bin, err := exec.LookPath("git")
	if err != nil {
		out.err = "git is not on PATH"
		return out
	}

	ctx, cancel := context.WithTimeout(ctx, filesGitDeadline)
	defer cancel()

	root, err := runGit(ctx, bin, "-C", dir, "rev-parse", "--show-toplevel")
	if err != nil {
		// The overwhelmingly common cause is exactly what it says. Anything
		// rarer (an unreadable .git, a version too old) still leaves the user
		// with a working listing and a pointer at git.
		out.err = "not a git worktree"
		return out
	}
	out.repoRoot = filepath.Clean(root)

	// A branch lookup that fails is NOT fatal: an unborn HEAD (a repo with no
	// commits yet) has no branch to name but does have a status worth showing.
	// Detached HEAD reports the literal "HEAD", which is not a branch name, so
	// it reports as no branch at all.
	if branch, err := runGit(ctx, bin, "-C", dir, "rev-parse", "--abbrev-ref", "HEAD"); err == nil && branch != "HEAD" {
		out.branch = branch
	}

	// -C the ROOT, not dir: porcelain paths are then unambiguously relative to
	// the root, whichever subdirectory was asked for.
	status, err := runGitRaw(ctx, bin, "-C", out.repoRoot, "status", "--porcelain=v1", "--untracked-files=all", "-z")
	if err != nil {
		out.err = "git status failed: " + gitErrText(err)
		return out
	}
	out.changed = parsePorcelainZ(out.repoRoot, status)
	return out
}

// runGit runs one git subcommand and returns its trimmed stdout.
func runGit(ctx context.Context, bin string, args ...string) (string, error) {
	out, err := runGitRaw(ctx, bin, args...)
	return strings.TrimSpace(out), err
}

// runGitRaw runs one git subcommand and returns stdout BYTE-FOR-BYTE.
//
// Output(), never CombinedOutput(): stderr interleaved into a NUL-separated
// record stream would be parsed as filenames. stdin is nil so a git that
// decides to prompt (a credential helper, say) hits EOF immediately instead of
// hanging until the deadline.
func runGitRaw(ctx context.Context, bin string, args ...string) (string, error) {
	cmd := exec.CommandContext(ctx, bin, args...)
	cmd.Stdin = nil
	out, err := cmd.Output()
	if err != nil {
		return "", err
	}
	return string(out), nil
}

// gitErrText renders a subprocess failure as one human sentence, preferring
// git's own words on stderr to Go's "exit status 128".
func gitErrText(err error) string {
	var ee *exec.ExitError
	if errors.As(err, &ee) {
		if msg := strings.TrimSpace(string(ee.Stderr)); msg != "" {
			if i := strings.IndexByte(msg, '\n'); i >= 0 {
				msg = msg[:i]
			}
			return msg
		}
	}
	return err.Error()
}

// parsePorcelainZ turns `status --porcelain=v1 -z` output into absolute path ->
// status token.
//
// The record is "XY<space>PATH" with NO trailing newline and no quoting: -z
// disables the quoting that would otherwise mangle a filename with a space,
// a quote or a newline in it, which is the reason this route pays for the NUL
// form instead of splitting lines.
//
// Renames are the one place the -z form differs from the human-readable one,
// and it is inverted from what the arrow suggests. Verified against a real
// repository:
//
//	human:  R  orig.txt -> renamed.txt
//	-z:     R  renamed.txt<NUL>orig.txt<NUL>
//
// so the record's OWN path is the DESTINATION and the extra field that follows
// is the ORIGIN. Consuming that extra field is not cosmetic: skip it and the
// origin is read as the next record's "XY path", and every status after the
// first rename is garbage.
//
// Only the destination is recorded. It is the path that exists on disk and
// therefore the only one that can appear in a listing, and for a copy (C) the
// origin genuinely has not changed -- claiming it did would be a lie about a
// file git says is clean.
func parsePorcelainZ(repoRoot, out string) map[string]string {
	changed := map[string]string{}
	fields := strings.Split(out, "\x00")
	for i := 0; i < len(fields); i++ {
		rec := fields[i]
		// "XY p" is the shortest legal record. This also swallows the empty
		// tail Split leaves after the final NUL.
		if len(rec) < 4 {
			continue
		}
		xy := rec[:2]
		if strings.ContainsAny(xy, "RC") {
			i++ // the origin path; see above
		}
		changed[filepath.Join(repoRoot, rec[3:])] = porcelainStatus(xy)
	}
	return changed
}

// porcelainStatus maps a porcelain XY pair onto the wire vocabulary.
//
// Order is the whole algorithm. Conflicts are tested before everything else
// because "AA" is both-added and "DD" is both-deleted -- reading either by its
// letters alone would report a merge conflict as an ordinary add or delete and
// send the user to commit it.
func porcelainStatus(xy string) string {
	switch {
	case xy == "??":
		return fileStatusUntracked
	case strings.ContainsRune(xy, 'U'), xy == "AA", xy == "DD":
		return fileStatusConflicted
	case strings.ContainsAny(xy, "RC"):
		return fileStatusRenamed
	case strings.ContainsRune(xy, 'D'):
		return fileStatusDeleted
	case strings.ContainsRune(xy, 'A'):
		return fileStatusAdded
	default:
		return fileStatusModified
	}
}
